### PR TITLE
Entity viewer sidebar analytics

### PR DESCRIPTION
**Pull request by imransl**
## Type

- [ ] Bug fix
- [x] New feature
- [ ] Improvement to existing feature

## Related JIRA Issue(s)
[ENSWBSITES-1367](https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-1367)

## Description
Add analytics to the sidebar of entity viewer. These were the four areas that are covered in this PR:

1. Clicking on sidebar tag
2. Clicking on external reference links in the Overview section
3. Clicking on external reference links in the External references section
4. Gene downloads in sidebar

~~TODO: Fix tests~~

## Deployment URL
http://ev-sidebar-analytics.review.ensembl.org/

## Views affected
Entity viewer -> Sidebar

### Other effects

- [x] I have checked any requirements for Google Analytics
- [ ] I have created any required tests
- [ ] The PR has as its final commit a WASM/JS update commit if any Rust files were updated.

### CODE REVIEW

File: src/ensembl/src/shared/components/instant-download/instant-download-gene/InstantDownloadGene.tsx [hunk 1]
```diff
@@ -117,8 +128,10 @@ const InstantDownloadGene = (props: Props) => {
 
     try {
       await fetchForGene(payload);
+      props.onDownloadSuccess?.(payload);
     } finally {
       resetCheckboxes();
+      props.onDownloadFailure?.(payload);
```
Comment (azangru) [inline src/ensembl/src/shared/components/instant-download/instant-download-gene/InstantDownloadGene.tsx:134]: the `finally` clause runs regardless of whether an error happened within `try`.

File: src/ensembl/src/content/app/entity-viewer/hooks/useEntityViewerAnalytics.ts [hunk 2]
```diff
@@ -157,14 +158,55 @@ const useEntityViewerAnalytics = () => {
...
     trackTabChange,
     trackFiltersPanelOpen,
     trackAppliedFilters,
     trackAppliedSorting,
     trackProteinInfoToggle,
     trackProteinDownload,
-    trackExternalLinkClickInProteinsList
+    trackExternalLinkClickInProteinsList,
+    trackXrefsTabSelection,
+    trackXrefLinkClick,
```
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/hooks/useEntityViewerAnalytics.ts:207]: Notice that a function above is called `trackExternalLinkClickInProteinsList`.

- at least within a single file (such as useEntityViewerAnalytics.ts) we should refer to the same concept by the same name: either ExternalLink or XrefLink
- these links exist in multiple places; so the name of the function should reflect where the link has been clicked.

(see also the dev branch here: https://github.com/Ensembl/ensembl-client/blob/dev/src/ensembl/src/content/app/entity-viewer/hooks/useEntityViewerAnalytics.ts#L211-L212)

File: src/ensembl/src/shared/components/instant-download/instant-download-gene/InstantDownloadGene.tsx (hunk 1 above)
Comment (imransl) [reply src/ensembl/src/shared/components/instant-download/instant-download-gene/InstantDownloadGene.tsx:134]: Thanks for noticing this! Should've gone into `catch`.

**Comment by imransl** [comment]
Just noticed that there are more conflicts after already fixing the conflicts 😁 I'll fix these and push again.

**Comment by imransl** [comment]
I have fixed the tests with @azangru's suggestion with the latest commit acd9c57

File: src/ensembl/src/content/app/entity-viewer/hooks/useEntityViewerAnalytics.ts [hunk 3]
```diff
@@ -117,13 +118,53 @@ const useEntityViewerAnalytics = () => {
     trackExternalLinkClick('gene_view_transcript_list', label);
```
[the remark this answers is not in the document]
Comment (imransl) [reply src/ensembl/src/content/app/entity-viewer/hooks/useEntityViewerAnalytics.ts:121]: I've made this change with the latest commit

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/gene-view-sidebar/external-references/GeneExternalReferences.tsx [hunk 4]
```diff
@@ -181,6 +199,7 @@ const GeneExternalReferences = () => {
 
 const TranscriptExternalReferencesGroups = (props: {
   transcript: Transcript;
+  externalReferenceClickHandler: (linkLabel: string) => void;
```
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/gene-view-sidebar/external-references/GeneExternalReferences.tsx:202]: `onClick`? `onReferenceClick`? From the viewpoint of this component, there's only one thing that one can click.

Comment (imransl) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/gene-view-sidebar/external-references/GeneExternalReferences.tsx]: ✅

File: src/ensembl/src/shared/components/view-in-app/ViewInApp.tsx [hunk 5]
```diff
@@ -85,14 +85,19 @@ export const ViewInApp = (props: ViewInAppProps) => {
...
+
+  if (onAnyAppClickFn) {
+    clickHandlers.push(() => onAnyAppClickFn(params.appName));
+  }
+
+  if (onAppClickFn) {
+    clickHandlers.push(() => onAppClickFn());
+  }
+
+  return clickHandlers ? () => clickHandlers.forEach((fn) => fn()) : undefined;
```
Comment (azangru) [inline src/ensembl/src/shared/components/view-in-app/ViewInApp.tsx:100]: 1) your `clickHandlers` are always truthy now.

Comment (imransl) [reply src/ensembl/src/shared/components/view-in-app/ViewInApp.tsx]: ✅

File: src/ensembl/src/content/app/entity-viewer/shared/components/entity-viewer-sidebar/entity-viewer-sidebar-modal/modal-views/EntityViewerBookmarks.tsx [hunk 6]
```diff
@@ -40,6 +42,23 @@ type PreviouslyViewedLinksProps = {
...
+    entity: {
+      entity_id: string;
+      label: string | string[];
+    },
+    index: number
+  ) => {
+    const { entity_id, label } = entity;
+    trackPreviouslyViewedLinkClick({
+      geneId: entity_id,
+      geneSymbol: typeof label === 'string' ? label : label[0],
```
[the remark this answers is not in the document]
Comment (imransl) [reply src/ensembl/src/content/app/entity-viewer/shared/components/entity-viewer-sidebar/entity-viewer-sidebar-modal/modal-views/EntityViewerBookmarks.tsx]: ✅

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/gene-view-sidebar/overview/GeneOverview.test.tsx [hunk 7]
```diff
@@ -34,6 +34,13 @@ jest.mock('../publications/GenePublications', () => () => (
   <div className="genePublications" />
 ));
 
+jest.mock(
+  'src/content/app/entity-viewer/hooks/useEntityViewerAnalytics',
+  () => () => ({
+    trackXrefLinkClick: jest.fn()
```
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/gene-view-sidebar/overview/GeneOverview.test.tsx:40]: Where is this function used?

Comment (imransl) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/gene-view-sidebar/overview/GeneOverview.test.tsx]: ✅

**Review by azangru — APPROVED** [review]

**Pull request merged**